batch/batch/driver/job.py

```diff
@@ -83,12 +83,14 @@ async def add_attempt_resources(db, batch_id, job_id, attempt_id, resources):
             # This must be sorted in order to match the order of values in the actual SQL table!
             _resources = dict(sorted(_resources.items()))
 
-            resource_args = [(batch_id, job_id, attempt_id, name, quantity) for name, quantity in _resources.items()]
+            resource_args = [(batch_id, job_id, attempt_id, quantity, name) for name, quantity in _resources.items()]
 
             await db.execute_many(
                 '''
-INSERT INTO `attempt_resources` (batch_id, job_id, attempt_id, resource, quantity)
-VALUES (%s, %s, %s, %s, %s)
+INSERT INTO `attempt_resources` (batch_id, job_id, attempt_id, resource_id, quantity)
```

Comment:
@danking Can you double check this query for me? I looked in the database and everything seems to be inserted properly (each resource has the right quantity), but I couldn't find documentation to make sure the execute many is handled correctly. I think it must be batching 5 individual insert statements?

Reply:
The join-insert defeats the bulk insert optimization, so this is just a loop over `execute`. That might slow down job insertion, but I kind of doubt this is the bottleneck of insertion. It can also be parallelized if necessary. Anyway, we should measure things like job insertion time!

https://github.com/PyMySQL/PyMySQL/blob/46d17402afaa07369b954eee026f68c5b96207ba/pymysql/cursors.py#L162-L197

Reply:
I do think that this non-trivially increases load on the driver as this is on the critical path for MJS/MJC. Can we simply load the resource -> resource id mapping into memory on the driver and remove the join? Since I don't believe we would ever alter a row in the resources table, that mapping shouldn't ever be stale, right?

Reply:
There's infrastructure to be able to do this in the BillingManagers where they have the resources in memory and update it from the database. Let's get #11840 merged in first before we add this though unless it's on the critical path.

Reply:
It's not urgent, I can hack around it for now